File: mobile/utils/specialityMatch.ts
const SPECIALITY_ALIASES: Record<string, string[]> = {
  cardiology: ["cardiol", "cardiac", "heart"],
  orthopedics: ["orthop", "orthopaedic", "bone", "ortho"],
  psychiatry: ["psychiat", "psych", "psychiatrist"],
  ophthalmology: ["ophthal", "eye", "optomet"],
  ent: ["otolaryng", "ear", "throat", "ent"],
  dentistry: ["dental", "dentist", "tooth", "oral"],
  "general medicine": ["general physician", "general medic", "physician", "gp", "internal"],
  gynecology: ["gynec", "gynaec", "obstetric", "obgyn"],
  dermatology: ["dermat", "skin"],
  pediatrics: ["pediatr", "paediatr", "child", "paediatric"],
  neurology: ["neurolog", "neuro", "brain"],
  gastroenterology: ["gastro", "gastroenterol", "digest"],
};

/** Match doctor specialization against a home speciality label or search token. */
export function matchesSpeciality(specialization: string, query: string): boolean {
  const spec = specialization.toLowerCase().trim();
  const target = query.toLowerCase().trim();
  if (!spec || !target) return false;
  if (spec.includes(target) || target.includes(spec)) return true;

  const aliases = SPECIALITY_ALIASES[target] ?? [];
  if (aliases.some((a) => spec.includes(a))) return true;

  for (const [key, words] of Object.entries(SPECIALITY_ALIASES)) {
    if (target.includes(key) || key.includes(target)) {
      if (words.some((w) => spec.includes(w)) || spec.includes(key)) return true;
    }
  }

  if (target.length >= 3) {
    for (const [key, words] of Object.entries(SPECIALITY_ALIASES)) {
      if (target.includes(key.slice(0, 4)) || words.some((w) => target.includes(w.slice(0, 4)))) {
        if (spec.includes(key) || words.some((w) => spec.includes(w))) return true;
      }
    }
  }

  return false;
}

File: flutter_mobile/lib/utils/speciality_match.dart
/// Port of mobile/utils/specialityMatch.ts
const _aliases = <String, List<String>>{
  'cardiology': ['cardiol', 'cardiac', 'heart', 'cardiologist'],
  'orthopedics': ['orthop', 'orthopaedic', 'bone', 'ortho', 'orthoped'],
  'psychiatry': ['psychiat', 'psych', 'psychiatrist'],
  'ophthalmology': ['ophthal', 'eye', 'optomet', 'ophthalmologist'],
  'ent': ['otolaryng', 'ear', 'throat', 'ent'],
  'dentistry': ['dental', 'dentist', 'tooth', 'oral'],
  'general medicine': ['general physician', 'general medic', 'physician', 'gp', 'internal'],
  'gynecology': ['gynec', 'gynaec', 'obstetric', 'obgyn', 'gynecologist'],
  'dermatology': ['dermat', 'skin', 'dermatologist'],
  'pediatrics': ['pediatr', 'paediatr', 'child', 'paediatric', 'pediatrician', 'pediatricians'],
  'neurology': ['neurolog', 'neuro', 'brain', 'neurologist'],
  'gastroenterology': ['gastro', 'gastroenterol', 'digest', 'gastroenterologist'],
};

bool matchesSpeciality(String specialization, String query) {
  final spec = specialization.toLowerCase().trim();
  final target = query.toLowerCase().trim();
  if (spec.isEmpty || target.isEmpty) return false;
  if (spec.contains(target) || target.contains(spec)) return true;

  final aliases = _aliases[target];
  if (aliases != null && aliases.any((a) => spec.contains(a))) return true;

  for (final entry in _aliases.entries) {
    final key = entry.key;
    if (target.contains(key) || key.contains(target)) {
      if (entry.value.any((w) => spec.contains(w)) || spec.contains(key)) return true;
    }
  }

  return false;
}

/// Resolve home/dashboard label to canonical speciality key for strict filtering.
String? canonicalSpecialityKey(String query) {
  final target = query.toLowerCase().trim();
  if (target.isEmpty) return null;
  if (_aliases.containsKey(target)) return target;
  for (final key in _aliases.keys) {
    if (target.contains(key) || key.contains(target)) return key;
    if (_aliases[key]!.any((a) => target.contains(a))) return key;
  }
  return null;
}

String formatSpecialityTitle(String filterKey) {
  final s = filterKey.trim();
  if (s.isEmpty) return 'Doctors';
  if (s.length == 1) return '${s.toUpperCase()} Doctors';
  return '${s[0].toUpperCase()}${s.substring(1)} Doctors';
}
